File: ios/native/MauiEmarsys/DotnetEmarsysInApp.swift
//
//  Copyright © 2024 Emarsys. All rights reserved.
//

import Foundation
import EmarsysSDK

@objc(DotnetEmarsysInApp)
public class DotnetEmarsysInApp: NSObject {
    
    @objc
    public static func setEventHandler(_ eventHandler: @escaping EventHandler) {
        Emarsys.inApp.eventHandler = eventHandler
    }
    
    @objc
    public static func pause() {
        Emarsys.inApp.pause()
    }
    
    @objc
    public static func resume() {
        Emarsys.inApp.resume()
    }
    
    @objc
    public static func isPaused() -> Bool {
        return Emarsys.inApp.isPaused()
    }
    
    @objc
    public static func createInlineInAppView() -> UIView {
        return EMSInlineInAppView()
    }
    
    @objc
    public static func setInlineInAppEventHandler(_ view: UIView, _ eventHandler: @escaping EventHandler) {
        (view as? EMSInlineInAppView)?.eventHandler = eventHandler
    }
    
    @objc
    public static func setInlineInAppCompletionListener(_ view: UIView, _ completionBlock: @escaping CompletionBlock) {
        (view as? EMSInlineInAppView)?.completionBlock = completionBlock
    }
    
    @objc
    public static func setInlineInAppCloseListener(_ view: UIView, _ closeBlock: @escaping () -> Void) {
        (view as? EMSInlineInAppView)?.closeBlock = closeBlock
    }
    
    @objc
    public static func loadInlineInApp(_ view: UIView, _ viewId: String) {
        (view as? EMSInlineInAppView)?.loadInApp(viewId: viewId)
    }
    
    @objc
    public static func setOnEventActionEventHandler(_ eventHandler: @escaping EventHandler) {
        Emarsys.onEventAction.eventHandler = eventHandler
    }
    
}

File: android/native/emarsys/src/main/java/com/emarsys/maui/DotnetEmarsysInApp.java
package com.emarsys.maui;

import android.content.Context;
import android.view.View;
import androidx.annotation.NonNull;
import com.emarsys.Emarsys;
import com.emarsys.inapp.ui.InlineInAppView;
import org.json.JSONObject;
import kotlin.Unit;

public class DotnetEmarsysInApp {

    public static void setEventHandler(@NonNull EventHandler eventHandler) {
        Emarsys.getInApp().setEventHandler(eventHandler::handleEvent);
    }

    public static void pause() {
        Emarsys.getInApp().pause();
    }

    public static void resume() {
        Emarsys.getInApp().resume();
    }

    public static boolean isPaused() {
        return Emarsys.getInApp().isPaused();
    }

    public static @NonNull View createInlineInAppView(@NonNull Context context) {
        return new InlineInAppView(context);
    }

    public interface InlineInAppEventHandler {
        void handleEvent(String eventName, JSONObject payload);
    }

    public static void setInlineInAppEventHandler(@NonNull View view, @NonNull InlineInAppEventHandler eventHandler) {
        ((InlineInAppView) view).setOnAppEventListener((eventName, json) -> {
            eventHandler.handleEvent(eventName, json.optJSONObject("payload"));
            return Unit.INSTANCE;
        });
    }

    public static void setInlineInAppCompletionListener(@NonNull View view, @NonNull CompletionListener completionListener) {
        ((InlineInAppView) view).setOnCompletionListener(completionListener::onCompleted);
    }

    public interface InlineInAppCloseListener {
        void onClosed();
    }

    public static void setInlineInAppCloseListener(@NonNull View view, @NonNull InlineInAppCloseListener closeListener) {
        ((InlineInAppView) view).setOnCloseListener(() -> {
            closeListener.onClosed();
            return Unit.INSTANCE;
        });
    }

    public static void loadInlineInApp(@NonNull View view, @NonNull String viewId) {
        ((InlineInAppView) view).loadInApp(viewId);
    }

    public static void setOnEventActionEventHandler(@NonNull EventHandler eventHandler) {
        Emarsys.getOnEventAction().setOnEventActionEventHandler(eventHandler::handleEvent);
    }

}
